### ISSUE TITLE
E2E tests on GKE are failing

### ISSUE THREAD

**Issue by barkbay** [elastic/cloud-on-k8s#3822]
E2E [tests on GKE are failing](https://devops-ci.elastic.co/job/cloud-on-k8s-e2e-tests-stack-versions/222/console). But it does not seem to be related to the tests themselves.

After after a few moments _(1 hour ?)_  there is this error in the logs:

```
[2020-10-09T00:17:27.834Z] 2020/10/09 00:17:21 Authenticating as service account...
[2020-10-09T00:17:27.834Z] 2020/10/09 00:17:21 Setting CLOUDSDK_CONFIG=/root/.gcp
[2020-10-09T00:17:27.834Z] Updated property [core/project].
....
[2020-10-09T00:30:20.154Z] {"Time":"2020-10-09T00:30:10.910703292Z","Action":"run","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples"}^M
[2020-10-09T00:30:20.154Z] {"Time":"2020-10-09T00:30:10.910872081Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples","Output":"=== RUN   TestSamples\n"}^M
[2020-10-09T00:30:20.154Z] {"Time":"2020-10-09T00:30:10.911512713Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples","Output":"{\"log.level\":\"info\",\"@timestamp\":\"2020-10-09T00:30:10.911Z\",\"log.logger\":\"e2e\",\"message\":\"Test context initialized\",\"service.version\":\"0.0.0-SNAPSHOT+00000000\",\"service.type\":\"eck\",\"ecs.version\":\"1.4.0\",\"context\":{\"operator\":{\"name\":\"e2e-i8i06-operator\",\"namespace\":\"e2e-i8i06-elastic-system\",\"operator_replicas\":1,\"managed_namespaces\":[\"e2e-i8i06-mercury\",\"e2e-i8i06-venus\"]},\"e2e_image\":\"docker.elastic.co/eck-ci/eck-e2e-tests:d2fe6c63\",\"e2e_namespace\":\"e2e-i8i06\",\"e2e_service_account\":\"e2e-i8i06\",\"elastic_stack_version\":\"7.10.0-SNAPSHOT\",\"log_verbosity\":1,\"operator_image\":\"docker.elastic.co/eck-snapshots/eck-operator-dev-ci:1.4.0-SNAPSHOT-d2fe6c63\",\"test_license\":\"/var/run/secrets/e2e/test-license.json\",\"test_license_pkey_path\":\"/var/run/secrets/e2e/dev-private.key\",\"test_regex\":\"^Test\",\"test_run\":\"e2e-i8i06\",\"monitoring_secrets\":\"/go/src/github.com/elastic/cloud-on-k8s/.ci/monitoring-secrets.json\",\"test_timeout\":300000000000,\"auto_port_forwarding\":false,\"deploy_chaos_job\":false,\"local\":false,\"ignore_webhook_f"}^M
[2020-10-09T00:30:20.154Z] {"Time":"2020-10-09T00:30:10.911564815Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples","Output":"ailures\":false,\"ocp_cluster\":false,\"ocp3_cluster\":false,\"pipeline\":\"e2e/stack-versions\",\"build_number\":\"160\",\"provider\":\"gke\",\"clusterName\":\"eck-7x-snapshot-160-e2e\",\"kubernetes_version\":\"default\"}}\n"}^M
[2020-10-09T00:30:20.154Z] {"Time":"2020-10-09T00:30:10.927314578Z","Action":"run","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana"}^M
[2020-10-09T00:30:20.154Z] {"Time":"2020-10-09T00:30:10.92735188Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana","Output":"=== RUN   TestSamples/apm-apm_es_kibana\n"}^M
[2020-10-09T00:30:20.154Z] {"Time":"2020-10-09T00:30:11.831233425Z","Action":"run","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/K8S_should_be_accessible"}^M
[2020-10-09T00:30:20.154Z] {"Time":"2020-10-09T00:30:11.831283548Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/K8S_should_be_accessible","Output":"=== RUN   TestSamples/apm-apm_es_kibana/K8S_should_be_accessible\n"}^M
[2020-10-09T00:30:20.154Z] {"Time":"2020-10-09T00:30:11.849907525Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/K8S_should_be_accessible","Output":"Retries (5m0s timeout): .\n"}^M
[2020-10-09T00:30:20.154Z] {"Time":"2020-10-09T00:30:11.849964489Z","Action":"run","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/Label_test_pods"}^M
[2020-10-09T00:30:20.154Z] {"Time":"2020-10-09T00:30:11.849986009Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/Label_test_pods","Output":"=== RUN   TestSamples/apm-apm_es_kibana/Label_test_pods\n"}^M
[2020-10-09T00:30:20.154Z] {"Time":"2020-10-09T00:30:11.903594051Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/Label_test_pods","Output":"Retries (5m0s timeout): .\n"}^M
[2020-10-09T00:30:20.154Z] {"Time":"2020-10-09T00:30:11.903643796Z","Action":"run","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/Elasticsearch_CRDs_should_exist"}^M
[2020-10-09T00:30:20.154Z] {"Time":"2020-10-09T00:30:11.903653646Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/Elasticsearch_CRDs_should_exist","Output":"=== RUN   TestSamples/apm-apm_es_kibana/Elasticsearch_CRDs_should_exist\n"}^M
[2020-10-09T00:30:20.155Z] {"Time":"2020-10-09T00:30:11.924425899Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/Elasticsearch_CRDs_should_exist","Output":"Retries (5m0s timeout): .\n"}^M
[2020-10-09T00:30:20.155Z] {"Time":"2020-10-09T00:30:11.924464625Z","Action":"run","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/Webhook_endpoint_should_not_be_empty"}^M
[2020-10-09T00:30:20.155Z] {"Time":"2020-10-09T00:30:11.924469949Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/Webhook_endpoint_should_not_be_empty","Output":"=== RUN   TestSamples/apm-apm_es_kibana/Webhook_endpoint_should_not_be_empty\n"}^M
[2020-10-09T00:30:20.155Z] {"Time":"2020-10-09T00:30:11.928961066Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/Webhook_endpoint_should_not_be_empty","Output":"Retries (5m0s timeout): .\n"}^M
[2020-10-09T00:30:20.155Z] {"Time":"2020-10-09T00:30:11.929000271Z","Action":"run","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/Remove_Elasticsearch_if_it_already_exists"}^M
[2020-10-09T00:30:20.155Z] {"Time":"2020-10-09T00:30:11.929027891Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/Remove_Elasticsearch_if_it_already_exists","Output":"=== RUN   TestSamples/apm-apm_es_kibana/Remove_Elasticsearch_if_it_already_exists\n"}^M
[2020-10-09T00:30:20.155Z] {"Time":"2020-10-09T00:30:11.940501739Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/Remove_Elasticsearch_if_it_already_exists","Output":"Retries (5m0s timeout): .\n"}^M
[2020-10-09T00:30:20.155Z] {"Time":"2020-10-09T00:30:11.940535251Z","Action":"run","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/K8S_should_be_accessible#01"}^M
[2020-10-09T00:30:20.155Z] {"Time":"2020-10-09T00:30:11.940544468Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e","Test":"TestSamples/apm-apm_es_kibana/K8S_should_be_accessible#01","Output":"=== RUN   TestSamples/apm-apm_es_kibana/K8S_should_be_accessible#01\n"}^M
... 1 hour later ? ...
[2020-10-09T01:19:43.643Z] {"Time":"2020-10-09T01:19:43.473206111Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/beat","Test":"TestPacketbeatConfig/Beat_should_be_created","Output":"=== RUN   TestPacketbeatConfig/Beat_should_be_created\n"}^M
[2020-10-09T01:19:43.643Z] {"Time":"2020-10-09T01:19:43.478054Z","Action":"run","Package":"github.com/elastic/cloud-on-k8s/test/e2e/beat","Test":"TestPacketbeatConfig/ES_certificate_authority_should_be_set_and_deployed"}^M
[2020-10-09T01:19:43.644Z] {"Time":"2020-10-09T01:19:43.478086252Z","Action":"output","Package":"github.com/elastic/cloud-on-k8s/test/e2e/beat","Test":"TestPacketbeatConfig/ES_certificate_authority_should_be_set_and_deployed","Output":"=== RUN   TestPacketbeatConfig/ES_certificate_authority_should_be_set_and_deployed\n"}^M
[2020-10-09T01:19:46.185Z] E1009 01:19:46.011020   15958 reflector.go:383] pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125: Failed to watch *v1.Event: Get "https://34.89.112.117/api/v1/events?allowWatchBookmarks=true&resourceVersion=2491&timeoutSeconds=581&watch=true": error executing access token command "/usr/local/google-cloud-sdk/bin/gcloud config config-helper --format=json": err=exit status 1 output= stderr=ERROR: (gcloud.config.config-helper) You do not currently have an active account selected.
```

This error repeats until e2e test timeout (10h)

Could be related to https://github.com/elastic/cloud-on-k8s/pull/3771

**Comment by david-kow** [comment]
I think the issue is that credentials have a token expiry set to 1 hour into the future. When running locally, I see the same thing:

```
2020/10/12 07:51:28 Authenticating as service account...
2020/10/12 07:51:30 Setting CLOUDSDK_CONFIG=/root/.gcp
Updated property [core/project].
Activated service account credentials for: [k8s-operator-ci@elastic-cloud-dev.iam.gserviceaccount.com]
```

And one of the log files:
```
... 
   "token_expiry": "2020-10-12T08:51:30Z"
...
```

**Comment by thbkrkr** [comment]
In #3771, I made reusable the function to perform authentication on GKE (https://github.com/elastic/cloud-on-k8s/pull/3771/commits/7b7510d43ab4c2597cd016947603fd1ab8e34242). 
This has notably changed the fact that we now always: 
- write the service account file used to execute `gcloud auth activate-service-account --key-file=...` in a different place (`~/.gcp`)
- set `CLOUDSDK_CONFIG=~/.gcp`
- and set the project gcloud (`gcloud config set project %s`)

It seemed to me that it would have no impact but maybe not actually?

**Comment by barkbay** [comment]
> * set `CLOUDSDK_CONFIG=~/.gcp`

I tried to do some tests, `CLOUDSDK_CONFIG` does not seem to behave exactly as with the default configuration directory. For example I'm not able to print my access token with `gcloud auth application-default print-access-token`

I'm wondering if we should also set `GOOGLE_APPLICATION_CREDENTIALS` as described [here](https://cloud.google.com/docs/authentication/getting-started#setting_the_environment_variable).

**Comment by barkbay** [comment]
> I'm wondering if we should also set `GOOGLE_APPLICATION_CREDENTIALS` as described [here](https://cloud.google.com/docs/authentication/getting-started#setting_the_environment_variable).

I'm doing a test locally, I will update this issue with my findings.

**Comment by david-kow** [comment]
I think I know what's happening.

**tldr:** 
We only set `CLOUDSDK_CONFIG` for the deployer, but not when it's needed later during e2e run to find the right config it's not set.

**Longer version:**
When deployer sets up the cluster we set the said env variable and fetch the credentials. The credentials are set to expire after an hour. During that time everything works and tests are being run. When hour passes and the token is expired, `kubectl` calls `/usr/local/google-cloud-sdk/bin/gcloud  config config-helper --format=json` as specified by the `kubectl` config generated by `gcloud auth activate-service-account`:
```
...
    auth-provider:
      config:
        cmd-args: config config-helper --format=json
        cmd-path: /usr/local/google-cloud-sdk/bin/gcloud
        expiry: "2020-10-12T11:02:38Z"
...
```

As this is called without `CLOUDSDK_CONFIG` set, the call fails with the error from the issue description:
```
error executing access token command "/usr/local/google-cloud-sdk/bin/gcloud config config-helper --format=json": err=exit status 1 output= stderr=ERROR: (gcloud.config.config-helper) You do not currently have an active account selected.
```

**Comment by thbkrkr** [comment]
We could just stop using a non default cloud SDK path for GKE, as it was before adding OCP3 support?

```diff
diff --git a/hack/deployer/runner/gke.go b/hack/deployer/runner/gke.go
index 195ee0bf..9406e236 100644
--- a/hack/deployer/runner/gke.go
+++ b/hack/deployer/runner/gke.go
@@ -80,7 +80,7 @@ func (gdf *GkeDriverFactory) Create(plan Plan) (Driver, error) {
 func (d *GkeDriver) Execute() error {
        if err := authToGCP(
                d.plan.VaultInfo, GkeVaultPath, GkeServiceAccountVaultFieldName,
-               d.plan.ServiceAccount, true, false, d.ctx["GCloudProject"],
+               d.plan.ServiceAccount, false, false, d.ctx["GCloudProject"],
        ); err != nil {
                return err
        }
```

**Comment by barkbay** [comment]
Closed by #3827
